resources/js/mp-history/game.tsx

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import StringWithComponent from 'components/string-with-component';
+import BeatmapJson from 'interfaces/beatmap-json';
+import BeatmapsetJson from 'interfaces/beatmapset-json';
+import LegacyMatchEventJson from 'interfaces/legacy-match-event-json';
+import Ruleset from 'interfaces/ruleset';
+import ScoreJson from 'interfaces/score-json';
+import UserJson from 'interfaces/user-json';
+import _ from 'lodash';
+import * as React from 'react';
+import { classWithModifiers } from 'utils/css';
+import { formatNumber } from 'utils/html';
+import { trans, transExists } from 'utils/lang';
+import GameHeader from './game-header';
+import Score from './score';
+
+interface TeamScores {
+  blue: number;
+  red: number;
+}
+
+type SortedScore = ScoreJson & {
+  teamRank: number;
+};
+
+interface Props {
+  event: LegacyMatchEventJson;
+  teamScores: TeamScores;
+  users: Partial<Record<number, UserJson>>;
+}
+
+export default function Game(props: Props) {
+  const game = props.event.game!;
+
+  if (game.scores.some((score) => score.match == null)) {
+    throw new Error('scores are missing match data');
+  }
```

Comment:
`!` is generally avoided and I think `scores` field should have match specific type instead.

Reply:
> should have match specific type instead.

Not sure where you want to have that type live. I put it next to `LegacyMatchGameJson` in https://github.com/ppy/osu-web/pull/12081/commits/1c2bd50bf2116ef7e683da383381379b1cf69df3, but that required an extra export that may be considered a touch ugly.

Reply:
I think it's usually (or sometimes) in the same file as the base json (score-json in this case)